src/@polymer/lit-element.ts

```diff
@@ -12,53 +12,167 @@
+   * @param {*} _prevProps Previous element properties
+   * @returns {boolean} Default implementation always returns true.
+   */
+  protected _shouldPropertiesChange(_props: any, _changedProps: any, _prevProps: any) {
+    return true;
+  }
+
+  /**
+   * Override which always calls `render` and `didRender` to perform
+   * element rendering.
+   * @param {*} props Current element properties
+   * @param {*} changedProps Changing element properties
+   * @param {*} prevProps Previous element properties
+   */
+  protected _propertiesChanged(props: any, changedProps: any, prevProps: any) {
+    this.__isInvalid = false;
+    super._propertiesChanged(props, changedProps, prevProps);
+    const result = this.render(props);
     if (result) {
       render(result, this.shadowRoot!);
     }
-    if (this._nextRenderedResolver) {
-      this._nextRenderedResolver();
-      this._nextRenderedResolver = null;
-      this._nextRendered = null;
+    this.didRender(props, changedProps, prevProps);
+    if (this.__resolveRenderComplete) {
+      this.__resolveRenderComplete();
+    }
+  }
+
+  protected _flushProperties() {
+    super._flushProperties();
+    if (this.__dataPending) {
+      console.warn(`Setting properties in response to properties changing considered harmful. Offending properties: ${Object.keys(this.__dataPending)}.`);
     }
   }
 
   /**
-   * Return a template result to render using lit-html.
+   * Returns a lit-html TemplateResult which is rendered into the element's
+   * shadowRoot. This method must be implemented.
+   * @param {*} _props Current element properties
+   * @returns {TemplateResult} Must return a lit-html TemplateResult.
    */
-  render(_props: object): TemplateResult {
+  protected render(_props: any): TemplateResult {
     throw new Error('render() not implemented');
   }
 
+  /**
+   * Called after element dom has been rendered. Implement to
+   * directly access element DOM.
+   * @param {*} _props Current element properties
+   * @param {*} _changedProps Changing element properties
+   * @param {*} _prevProps Previous element properties
+   */
+  protected didRender(_props: any, _changedProps: any, _prevProps: any) {}
+
+  /**
+   * Provokes the element to asynchronously re-render.
+   */
   invalidate() {
     this._invalidateProperties();
   }
 
-  get nextRendered() {
-    if (!this._nextRendered) {
+  /**
+   * Override which provides tracking of invalidated state.
+  */
+  _invalidateProperties() {
+    this.__isInvalid = true;
+    super._invalidateProperties();
+  }
+
+  /**
+   * Returns a promise which resolves after the element next renders.
+   */
+  get renderComplete() {
+    if (!this.__renderComplete) {
       // TODO(sorvell): handle rejected render.
-      this._nextRendered = new Promise((resolve) => {
-        this._nextRenderedResolver = resolve;
+      this.__renderComplete = new Promise((resolve) => {
+        this.__resolveRenderComplete = resolve;
```

Comment:
Rather than chaining off of the promise to clear the resolve function & promise, it could be done by wrapping resolve here right?

```js
this.__resolveRenderComplete = () => {
  resolve();
  this.__renderComplete = this.__resolveRenderComplete = null;
}
```
I guess it does a similar thing since `this.__renderComplete.then(...` is the first thing chained off the fresh promise and guaranteed to go first?

The async'ness is hard to reason about, i.e. whether there's a hole of time where someone could request the promise after it's been resolved, before it's been reset...

Reply:
PTAL